### ISSUE TITLE
Problem after first update

### ISSUE THREAD

**Issue by NeoMorfeo** [ad-ha/ide_api#3]
After first configuration and first update of the value, i got this error on the consecuents calls...

Let me know if you need more information

PS: great job, please, let me know if you need help developing or debuggin!

![image](https://user-images.githubusercontent.com/3766333/148059415-cf44c141-8bd6-44e8-ad0e-60305ed0df4b.png)

```
Logger: homeassistant.helpers.entity
Source: custom_components/ide/ide_api.py:160
Integration: i-DE Energy Monitor (documentation, issues)
First occurred: 12:54:11 (1 occurrences)
Last logged: 12:54:11

Update for sensor.ide_meter_reading fails
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 487, in async_update_ha_state
    await self.async_device_update()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 699, in async_device_update
    await task
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/ide/sensor.py", line 149, in update
    meter = ides.watthourmeter()
  File "/config/custom_components/ide/ide_api.py", line 169, in watthourmeter
    return self.measurement()["meter"]
  File "/config/custom_components/ide/ide_api.py", line 160, in measurement
    "id": json_response["codSolicitudTGT"],
KeyError: 'codSolicitudTGT'
```

**Comment by ad-ha** [comment]
Another way to try it:

Login to your i-de account as you do normally on a web browser.

After login, entre this address on the browser: https://www.i-de.es/consumidores/rest/escenarioNew/obtenerMedicionOnline/24

If you get a WU1, the issue will be with the login or i-DE response, not the Integration. If you get something similar to the one below, it should all be working fine on HA:

`{"valMagnitud":"470.0","valInterruptor":"1","valEstado":"09","valLecturaContador":"24353","codSolicitudTGT":"010xxxxx9600"}`

**Comment by NeoMorfeo** [comment]
After one hour more, it looks working again, I will try to monitor it a few days, and know if there is something puntual or its an issue.

I have checked the method you describe, open the URL, login and navigate to the other URL, and it takes long time without response (timeout). After a few tries, it gave me an error "Se ha producido un error. No se puede completar la operación solicitada." With status code 403 (and I was logged out on the web) Maybe we are reaching limits?

**Comment by NeoMorfeo** [comment]
After more or less one day with that, I notice a few errors/warnings, maybe need further investigation, but the information obtained is not consistent referring to the time when it could be obtained:

Yesterday (04/01/2022) measurements  :
![image](https://user-images.githubusercontent.com/3766333/148180639-7f7a7450-1986-4981-8f06-396572ede38e.png)


Today (05/01/2022) measurements :
![image](https://user-images.githubusercontent.com/3766333/148180468-a3e649c2-8987-4a4b-be8d-24bcce229a4d.png)

And the logs:
![image](https://user-images.githubusercontent.com/3766333/148180708-f5d03e79-2324-4279-9d6a-772531291493.png)

![image](https://user-images.githubusercontent.com/3766333/148180770-0d871e2a-bdb7-452c-8f2c-db2d0f2f6b61.png)


Today I have tested the url steps you say yesterday, and now it works as you said.. but something missed here.

Please let me know if you need more information.

**Comment by ad-ha** [comment]
Hi @NeoMorfeo,

Which version of the integration are you using? 

I do not understand what you mean by inconsistency between both days on the energy panel. Can you elaborate, please.

The error that you are getting on the log is related to the i-DE response, not the integration itself. You have to keep in mind that each reading goes straight to your home meter and asks for the actual consumption and cumulative reading; that's why it takes so long to get a response from the server. If something fails in between, it might be due to your meter or the i-DE server.

Cheers

**Comment by ad-ha** [comment]
Btw, just wondering on things that might impact the response from i-DE.

What kind of profile do you have on i-DE? Have a look here: https://www.i-de.es/consumidores/web/home/personal-area/userData

Is "Perfil Usuario" Usuario Avanzado?

**Comment by NeoMorfeo** [comment]
I have installed it using HACS, and there I have no information about the version (and don't know how to obtain it)... sorry

I know the operation of measure is not controlled by the integration, just sharing just in case you could infer more info on that.

To explain me more about the consistence, I don't know in which time or when the integration ask for the measures but, i would be nice to have the same period always (every 1 hour, every 2), when I said the data has inconsistences, I was talking about that, I think would be better to have that data every period, because as you can see on my previous images, some time it takes the measurement after one hour, and sometimes only after 3 or 4. 

Maybe is something I don't know read properly on the data?

**Comment by NeoMorfeo** [comment]
> Btw, just wondering on things that might impact the response from i-DE.
> 
> What kind of profile do you have on i-DE? Have a look here: https://www.i-de.es/consumidores/web/home/personal-area/userData
> 
> Is "Perfil Usuario" Usuario Avanzado?

I have "Usuario Avanzado"

**Comment by ad-ha** [comment]
@NeoMorfeo 

A ver si me explico. El sensor actualiza la lectura cada hora, pero puede que haya veces en que el contador no contesta y no se genera lectura o bien, esto ya más a menudo, hay veces en que también la web de i-DE tiene problemas para reportar. La inconsistencia de tus valores podría ser por eso.

A mi me pasa lo mismo:

![image](https://user-images.githubusercontent.com/59612788/148250046-b03d26f3-8c64-4357-a2ec-8aa902731e30.png)

**Comment by NeoMorfeo** [comment]
> @NeoMorfeo
> 
> A ver si me explico. El sensor actualiza la lectura cada hora, pero puede que haya veces en que el contador no contesta y no se genera lectura o bien, esto ya más a menudo, hay veces en que también la web de i-DE tiene problemas para reportar. La inconsistencia de tus valores podría ser por eso.
> 
> A mi me pasa lo mismo:
> 
> ![image](https://user-images.githubusercontent.com/59612788/148250046-b03d26f3-8c64-4357-a2ec-8aa902731e30.png)

Perfecto, a eso me refería :) siendo ese el tema del muestreo del contador, poco se puede hacer, aunque me pregunto si no habrá almacenado en la API de I-DE ese dato, para consultarlo de ellos directo, en vez de la puntual del contador (no sé si es factible o si existe, solo pregunto 😄 )

**Comment by ad-ha** [comment]
Buenas,

He estado dando vueltas a esto y ya sé qué es lo que pasa y por qué aparecen esas "incongruencias" de lecturas.

Básicamente, la lectura del contador sólo devuelve número entero, sin decimales. Con lo cual, si a cada hora la lectura del contador no ha subido más de 1 kWh, el sensor no detecta cambios y da lectura 0. Por eso hay horas en que no aparece nada. Luego cuando detecta una variación superior a 1kWh, respecto a la última lectura, es cuando apunta el consumo.

@NeoMorfeo en tús gráficas es lo que pasa. Que cada intervalo en cero luego viene precedido de una lectura de 1 kWh.

Espero haber aclarado la cuestión.

Cierro el tópico
